### ISSUE TITLE
Unloading rJava and/or restarting JVM

### ISSUE THREAD

**Issue by jeroen** [s-u/rJava#25]
[Repost from [this SO question](https://stackoverflow.com/questions/24337383/unloading-rjava-and-or-restarting-jvm)]. 

I would like to use `rJava` in combination with `mcparallel` but obviously [the JVM cannot be forked](https://stat.ethz.ch/pipermail/r-devel/2013-November/067960.html). Therefore a separate JVM instance needs to be initiated for each child process, e.g:

```
library(rJava)
library(parallel)
myfile <-  system.file("tests", "test_import.xlsx", package = "xlsx")

#This works:
mccollect(mcparallel({
  #Automatically initiates JVM in child
  xlsx::read.xlsx(myfile, 1)
}))
```

However the problem in my case is that the JVM has already been initiated in the (main) parent process as well. This makes it impossible to use `rJava` in the child process:

```
#init JVM in parent
.jinit()

#Doesn't work anymore
mccollect(mcparallel({
  xlsx::read.xlsx(myfile, 1)
}))
```

 So what I really need is a way to shutdown/kill and restart the JVM in the child process. Simply  `detach("package:rJava", unload = TRUE)` doesn't seem to do the trick. The `force.init` parameter doesn't seem to result in a restart either:

```
#Also doesn't work:
.jinit()
mccollect(mcparallel({
  .jinit(force.init = TRUE)
  xlsx::read.xlsx(myfile, 1)
}))
```

Is there some way I can forcefully shutdown/kill the JVM in order to reinitiate it in the child process?

**Comment by ManuelB** [comment]
Hi,
just another bug report. I use RJDBC to connect to an SQL Server. When I requests a big table (1.400.000 rows). I get:

```
Error in .jcall(rp, "I", "fetch", stride, block) : 
  java.lang.OutOfMemoryError: Java heap space
```

It would be nice if it would be possible to set the max heap size:

```
options( java.parameters = "-Xmx2g" )
```

And and afterwards restart the JVM.

As a workaround I can just restart the whole rsession in RStudio.

/Manuel

**Comment by s-u** [comment]
Yes, you should really set `java.parameters` in your startup scripts such that it is always active. That said, I think you don't actually need it for RJDBC - you only need to run `gc()` between the `fetch()` calls so Java can release the data you already got on the R side.

**Comment by drknexus** [comment]
If I read the conversation up until now correctly, we are limited to one JVM per application and we can't guarantee/control a close of a JVM without generating overhead.  

Follow up naive question.  It is conceptually possible to somehow isolate the JVM on a separate process and then accomplish communications between that separate process and R via a socket?  A back of napkin script with DBI/RJDBC seemed to confirm that I could hold a separate db connection on a single SNOW node on localhost at the same time as the db connection on the master node.  Of course in that case the communication with the JVM happens entirely within the slave node rather than needing to access/interact with R objects on the master - so maybe there is no way to accomplish anything like this would a massive rewrite?

Of course, the socket would have to invalidate on a fork and spawn its own JVM - but that part seems trivial relative to the potential costs/issues of isolating the JVM on a separate process/in a separate application.  There would be overhead, of course, so this probably would have to be implemented as an option rather than as a default mode of operation.

**Comment by s-u** [comment]
Sure, it's relatively easy to have separate processes. However, the point of rJava is exactly to NOT have a separate process, otherwise you could have used Rserve, SNOW or anything like that which has other benefits - but speed and memory efficiency are not among them. So you should pick the tool that is best for the requirements you have.

**Comment by x1o** [comment]
To everyone reading this who have problems with the JVM not releasing the heap memory back to the OS: @asieira 's claim that

> the JVM grabs heap memory and never again returns it to the OS.

is true for only for the default GC in some early version of Java (prior to 1.8 IIRC). It is indeed possible to choose a GC that *would* release the heap on performing the full garbage collection, see https://www.javacodegeeks.com/2017/11/minimize-java-memory-usage-right-garbage-collector.html. Furthermore, some (modern) GC's do not even require a full garbage collection to keep the JVM's memory footprint down to the amount it actually uses, see https://openjdk.java.net/jeps/346 and https://wiki.openjdk.java.net/display/shenandoah/Main.

**Comment by s-u** [comment]
While trying to address #334 I have now final proof that it is impossible to truly unload or even re-initialize a VM. I have implemented a fully clean approach ([just pure C program](https://github.com/s-u/rJava/blob/958a7592db0303cce21e297caaaa036128ea7e46/src/djni.c)) which dynamically loads a JVM with `RTLD_LOCAL`, initializes it, calls `java.lang.System.getProperty("java.version")`, frees all references and then destroys the JVM with `DestroyJavaVM`. It turns out the JVM is never released even if the process which loaded it unloads `libjvm`. Surprisingly, the same is true if one extra level of indirection is inserted (a separate dynamic library loads JVM instead of the mains process - the library itself is unloaded, but `libjvm` is not). In addition, any attempts to re-create the JVM after using `DestroyJavaVM` fail with `JNI_CreateJavaVM` error. (All tests performed with LTS versions of OpenJDK Temurin). Hence I have to conclude that this is simply impossible with the current JVM implementations.

**Comment by chrisknoll** [comment]
Thanks for the update @s-u .  Not to hijack this thread but I do have a question related to JVM instantiation:

Will it ever be the case where JVM instances wil be shared across child R Session processes such that the rJava interface to calling out to the JVM host would amount to basically a new thread invocation on a shared VM across different R Sessions, or is it the case that as long as child R sessions materialize in separate processes, you will get a new JVM?

And taking this a little further, if ever R manages to introduce a way of parallel compute which doesn't involve new R processes (ie: work more single-process/multi-thread) would that mean that child 'sessions' of R would share a process and therefore the JVM?

I'm asking these so that we can avoid any pitfalls around some of our programming decisions about these technical considerations.
